app/src/main/java/com/example/filemanager/AllFilesViewModel.kt

```diff
@@ -1,21 +1,32 @@
 package com.example.filemanager
 
 import android.os.Environment
-import android.util.Log
 import androidx.lifecycle.MutableLiveData
 import androidx.lifecycle.ViewModel
-import java.io.File
 
 class AllFilesViewModel() : ViewModel() {
 
     val files = MutableLiveData<List<ListModel>>()
+    private val listRootDirectoryFiles: ListModel = ListModel(Environment.getExternalStorageDirectory())
+    var currentFile: ListModel? = listRootDirectoryFiles
 
     fun fetchFiles() {
-        val path = Environment.getExternalStorageDirectory().absolutePath ?: return
-        Log.d("PATH", path)
-        val files2 = File(path).listFiles()?.map {
+        files.postValue(listFiles(listRootDirectoryFiles))
+    }
+
+    fun listFiles(currentItem: ListModel): List<ListModel> {
+        return currentItem.file.listFiles()?.map {
             ListModel(it)
         } ?: emptyList()
-        files.postValue(files2)
+    }
+
+    fun listParentDirectory(): List<ListModel> {
+        val parentFile = currentFile?.file?.parentFile ?: return emptyList()
+        currentFile = ListModel(parentFile)
+        return listFiles(currentFile!!)
+    }
+
+    fun hasReachedRootFolder(): Boolean {
+        return currentFile!!.isRootDirectory()
```

Comment:
Aren't you afraid of using !! on a nullable value?

Reply:
I think no, because this variable will not be null